### PR TITLE
Implementation of RFC PR 37

### PR DESCRIPTION
**Pull request by kylef**
Implementation of RFC14 and RFC15.

- https://github.com/refractproject/rfcs/pull/37
- https://github.com/refractproject/rfcs/pull/36

### CODE REVIEW

File: refract-spec.md [hunk 1]
```diff
@@ -41,23 +41,17 @@ The Refract Element contains four properties: `element`, `meta`, `attributes`, a
...
-      - (array[Member Element])
+      - `id` - Unique Identifier, MUST be unique throughout the document
+      - `ref` (Element Pointer) - Pointer to referenced element or type
+      - `classes` (Array Element[String Element]) - Array of classifications for given element
+      - `title` (String Element) - Human-readable title of element
+      - `description` (String Element) - Human-readable description of element
+      - `links` (Array Element[Link Element]) - Meta links for a given element
 
-- `attributes` (enum)
+- `attributes` (object[string, Element])
```
Comment (pksunkara) [inline refract-spec.md]: What is the `[string, Element]`?

Comment (kylef) [reply refract-spec.md]: The keys may only be `string` and the values would only be `Element`. There are two generic components to an object. Do you think there is some better way to indicate this? We don't really have syntax for this in MSON.

Comment (pksunkara) [reply refract-spec.md]: They keys are always string in an object. We don't need to mention it. Just an `object` here would be enough. Regarding types of the values, you just need to mention `Element` for every property.

**Review by pksunkara — APPROVED** [review]

**Comment by smizell** [comment]
@kylef If I could make a small comment, is this change conflating the Refract spec and serialization? Should they be the same, or separate? 

For instance, conceptually, is the spec describing an object model or a serialization/representation model? I'm a little afraid this combines this altogether, though maybe that doesn't matter.

**Comment by pksunkara** [comment]
I think letting meta be an array of member elements even in the representation model is kinda confusing. Which is why we proposed to completely remove that part from the representation model.

**Comment by smizell** [comment]
Ah, I should clarify. I'm mostly referring to the values of meta properties, and how they are now elements instead of strings or arrays. I definitely agree with the issue with removing the ability to use arrays of elements.

I'm not sure this matters here, but I wonder if there should be a separation between the conceptual object model and the serialization format.

For example, should the conceptual object model say a meta title can be either a string or string element and the serialization rules says it's an element only? Is there anything lost in saying they are now only elements? Again not sure myself, so I'm curious what you all think.

**Comment by pksunkara** [comment]
> conceptual object model say a meta title can be either a string or string element and the serialization rules says it's an element only

I think this is what we are going for, but unfortunately, depicting this in the refract main spec might still be confusing for the first time consumers.

**Comment by smizell** [comment]
If that's what you are wanting, couldn't you:

1. Specify it explicitly in the spec that it can be a value or element
1. Add a note to clarify how values can be normal values or elements
1. Provide separate instructions for serialization and link to that in the clarifying note

Is the concern with confusion over the fact there is a conceptual model and serialization model? Or that the spec itself would be confusing that these values can be values or elements?

**Comment by pksunkara** [comment]
> Is the concern with confusion over the fact there is a conceptual model and serialisation model

I think this would be an issue. @kylef What do you think?

**Comment by kylef** [comment]
I've created https://github.com/refractproject/refract-spec/pull/63 which introduces a separate specification on how a Refract Element can be serialised as JSON.

In this PR, I have made changes so that the spec always use the canonical data type. Instead of things being either one or the other types, this spec shows a single one, the conceptual type and most-likely how it will be implemented in Refract libraries.

The separate PR #63 introduces a separate specification on how this relates to JSON Specifically. One concern I have is that the Refract specification does make heavy use of JSON examples and I was thinking that it may make sense to remove these to decouple JSON as much as possible from Refract spec.

Does this address your concerns @smizell?

**Comment by smizell** [comment]
Just to note, I don't want to dictate too much here. I just brought up that issue, especially given you all wanted to keep the option to do both values and elements. Do you think it makes sense to break up the conceptual model and serialization for this reason?

---

> One concern I have is that the Refract specification does make heavy use of JSON examples and I was thinking that it may make sense to remove these to decouple JSON as much as possible from Refract spec.

Sounds good to me. :)

**Pull request merged**